nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,520 @@
+  delete userStorage[userId];
+  res.redirect("/users");
+});
+```
+
+Don't forget to add the new route to your router!
+
+```javascript
+usersRouter.post("/:userId/delete", usersController.users_delete_post);
+```
+
+You'll notice we didn't need a `GET` route here since we're already redirecting to `/users`.
+
+We could go much deeper into working safely with forms, but we'll stop there. By now you can already see how helpful `express-validator` is, and how you can do almost anything you want with the `req.body` object.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+#### Add user details
+
+Expand the `User` model we created to include more details. Implement the following fields and validations to your model:
+
+- Email (required, must be formatted properly)
+- Age (optional, must be a number between 18 and 120)
+- Bio (optional, maximum 200 characters)
+
+Don't forget to update the view to display these new fields!
+
+#### Implement searching
+
+What if we want to search for a specific user in a list of thousands? We'll need a new route and view that lets clients search our list of users.
+
+1. Add a form (in `users.ejs` or another view) which accepts a `name` or `email` (or both!)
+1. Create a new route `/users/search` which accepts `GET` and `POST` requests.
+1. Add the search logic to your controller which searches your list for a matching user.
+   - Your `POST` request should handle searching for the user.
+   - Your `GET` request should then render the search result.
+1. Display the search results in a new view: `search.ejs`.
+
+</div>
+
+### Knowledge check
+
+The following questions are an opportunity to reflect on key topics in this lesson. If you can't answer a question, click on it to review the material, but keep in mind you are not expected to memorize or master this knowledge.
+
+- [What are the attributes of a form element and what do they do?](#html-forms-overview)
+- [How do you install and import express-validator in your project?](#validation-and-sanitization)
+- [How do you validate and sanitize form input using express-validator?](#understanding-the-body-function)
+- [What is the difference between validation and sanitization?](#validation-and-sanitization)
+- [How do you handle validation errors in Express routes?](#validation-results)
+- [What is the importance of escaping HTML characters in a form?](https://owasp.org/www-community/attacks/SQL_Injection)
+
+### Additional resources
+
+This section contains helpful links to related content. It isn't required, so consider it supplemental.
+
+- [Express request object](https://expressjs.com/en/api.html#req)
+- [Express response object](https://expressjs.com/en/api.html#res)
+  It's certainly not required viewing, but web dev simplified makes some of the best tutorials on Express. Take a watch if you need a refresher.
+- [Web dev simplified](https://youtu.be/SccSCuHhOw0?si=2dZ5Y4dvxyh7jpcy)
```

Comment:
Not sure if we need to include the complete video, do you have timestamp suggestion?

Reply:
Just want to tack on that the list item seems to have the incorrect formatting with split lines. We'd and also need a more descriptive link for accessibility, e.g. `Web Dev Simplified has a [tutorial on submitting forms and parsing form data in Express] if you want a refresher.` or whatever's most appropriate.

Reply:
I've updated the phrasing but left the full video in. I'll leave this open just in case we decide we want a timestamp, but this came from Nathaniel so I don't have a recommendation for a timestamp. For what it's worth the video has timestamps built into the video, so I don't think this is too big of an issue.